### PR TITLE
Support conditional copy-by-reference in local memory 

### PR DESCRIPTION
**Pull request by mairooni**
#### Description

This PR adds support for conditional copy-by-reference in local memory.

#### Problem description

If a plain copy-by-reference occurs in local memory, graal updates the references automatically. For example, the code below is legal and functionally correct: 

```java
public static void intPrivateCopyNoCondition(IntArray a, IntArray b) {
        for (@Parallel int i = 0; i < a.getSize(); i++) {
            int[] arrayA = new int[128];
            int[] arrayB = new int[128];
            for (int j = 0; j < 128; j++) {
                arrayA[j] = j;
                arrayB[j] = 2;
            }

            // copy arrayA to arrayB
            arrayB = arrayA;

            b.set(i, arrayB[i]);
        }
}
```
 However, if the copy takes place depending on a condition, e.g.: 
```java
 if ((a.get(i) % 2) == 0 ) {
       // if condition is true, copy arrayA to arrayB
        arrayB = arrayA;
 }
```
Tornado produces a compiler error. 
This PR adds the necessary functionality to make this conditional copy-by-reference legal. This patch is for OpenCL and PTX, SPIR-V needs further investigation. 

#### Backend/s tested

Mark the backends affected by this PR.

- [x] OpenCL
- [x] PTX
- [ ] SPIRV

#### OS tested

Mark the OS where this PR is tested.

- [x] Linux
- [ ] OSx
- [ ] Windows

#### Did you check on FPGAs?

If it is applicable, check your changes on FPGAs.

- [ ] Yes
- [x] No

#### How to test the new patch?

Run 
`tornado-test -V --fast uk.ac.manchester.tornado.unittests.arrays.TestArrayCopies`

### CODE REVIEW

**Comment by jjfumero** [comment]
[Review in progress] Test are passing for OpenCL and PTX.

File: tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/phases/TornadoFixedArrayCopyPhase.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+public class TornadoFixedArrayCopyPhase extends Phase {
+
+    @Override
+    public Optional<NotApplicable> notApplicableTo(GraphState graphState) {
+        return ALWAYS_APPLICABLE;
+    }
+
+    protected void run(StructuredGraph graph) {
+        for (ValuePhiNode phiNode : graph.getNodes().filter(ValuePhiNode.class)) {
+            if (phiNode.usages().filter(OffsetAddressNode.class).isNotEmpty() && phiNode.values().filter(FixedArrayNode.class).isNotEmpty()) {
```
Comment (jjfumero) [inline tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/phases/TornadoFixedArrayCopyPhase.java:51]: Add a method that returns true or false using these condition.

File: tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/phases/TornadoFixedArrayCopyPhase.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                ResolvedJavaType resolvedJavaType = fixedArrayNode.getElementType();
+                OCLArchitecture.OCLMemoryBase oclMemoryBase = fixedArrayNode.getMemoryRegister();
+                OffsetAddressNode offsetAddressNode = phiNode.usages().filter(OffsetAddressNode.class).first();
+                FixedArrayCopyNode fixedArrayCopyNode = new FixedArrayCopyNode(phiNode, resolvedJavaType, oclMemoryBase);
+                graph.addWithoutUnique(fixedArrayCopyNode);
+                offsetAddressNode.replaceFirstInput(phiNode, fixedArrayCopyNode);
+                // finally, since we know that the data accessed is a fixed array, fix the offset
+                ValuePhiNode privateIndex = getPrivateArrayIndex(offsetAddressNode.getOffset());
+                offsetAddressNode.setOffset(privateIndex);
+                break;
```
Comment (jjfumero) [inline tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/phases/TornadoFixedArrayCopyPhase.java:62]: Remove the break, we want to explore all copies within the kernel.

File: tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/phases/TornadoFixedArrayCopyPhase.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                FixedArrayNode fixedArrayNode = phiNode.values().filter(FixedArrayNode.class).first();
+                ResolvedJavaType resolvedJavaType = fixedArrayNode.getElementType();
+                OCLArchitecture.OCLMemoryBase oclMemoryBase = fixedArrayNode.getMemoryRegister();
+                OffsetAddressNode offsetAddressNode = phiNode.usages().filter(OffsetAddressNode.class).first();
+                FixedArrayCopyNode fixedArrayCopyNode = new FixedArrayCopyNode(phiNode, resolvedJavaType, oclMemoryBase);
+                graph.addWithoutUnique(fixedArrayCopyNode);
+                offsetAddressNode.replaceFirstInput(phiNode, fixedArrayCopyNode);
+                // finally, since we know that the data accessed is a fixed array, fix the offset
+                ValuePhiNode privateIndex = getPrivateArrayIndex(offsetAddressNode.getOffset());
+                offsetAddressNode.setOffset(privateIndex);
```
Comment (jjfumero) [inline tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/phases/TornadoFixedArrayCopyPhase.java:61]: Can `privateIndex` be null?

File: tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/graal/nodes/FixedArrayCopyNode.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        final Value result;
+        if (ptxKind.isF32() || ptxKind.isF64()) {
+            result = tool.newVariable(lirKind);
+            tool.append(new PTXLIRStmt.AssignStmt(result, new PTXUnary.Expr(PTXAssembler.PTXUnaryOp.CVT_FLOAT_RNE, lirKind, baseAddress)));
+            declarationPtr = new PTXBinary.Expr(pointerCopyTemplate, lirKind, result, baseAddress);
+        } else if (ptxKind.isS64()) {
+            result = tool.newVariable(lirKind);
+            tool.append(new PTXLIRStmt.AssignStmt(result, new PTXUnary.Expr(PTXAssembler.PTXUnaryOp.CVT_FLOAT, lirKind, baseAddress)));
+            declarationPtr = new PTXBinary.Expr(pointerCopyTemplate, lirKind, result, baseAddress);
+    } else {
```
Comment (jjfumero) [inline tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/graal/nodes/FixedArrayCopyNode.java:123]: Check the formatter

**Review by jjfumero — CHANGES_REQUESTED** [review]

**Comment by jjfumero** [comment]
Throw exception when this type of code is executed for the SPIR-V backend. This can be done within the a compilation phase.

**Comment by mairooni** [comment]
The comments have been addressed. @jjfumero please have another look.

**Review by jjfumero — APPROVED** [review]

**Comment by jjfumero** [comment]
Great. Thanks Mary!

File: tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/lir/OCLKind.java [hunk 5]
```diff
@@ -256,6 +256,44 @@ public static OCLAssembler.OCLBinaryTemplate resolvePrivateTemplateType(JavaKind
...
+            return OCLAssembler.OCLBinaryTemplate.PRIVATE_SHORT_ARRAY_PTR;
+        } else if (type == JavaKind.Long) {
+            return OCLAssembler.OCLBinaryTemplate.PRIVATE_LONG_ARRAY_PTR;
+        } else if (type == JavaKind.Char) {
+            return OCLAssembler.OCLBinaryTemplate.PRIVATE_CHAR_ARRAY_PTR;
+        } else if (type == JavaKind.Byte) {
+            return OCLAssembler.OCLBinaryTemplate.PRIVATE_BYTE_ARRAY_PTR;
+        }
+        return null;
+    }
```
Comment (mikepapadim) [inline tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/lir/OCLKind.java:276]:
```suggestion
public static OCLAssembler.OCLBinaryTemplate resolvePrivatePointerTemplate(JavaKind type) {
    return switch (type) {
        case Int -> OCLAssembler.OCLBinaryTemplate.PRIVATE_INT_ARRAY_PTR;
        case Double -> OCLAssembler.OCLBinaryTemplate.PRIVATE_DOUBLE_ARRAY_PTR;
        case Float -> OCLAssembler.OCLBinaryTemplate.PRIVATE_FLOAT_ARRAY_PTR;
        case Short -> OCLAssembler.OCLBinaryTemplate.PRIVATE_SHORT_ARRAY_PTR;
        case Long -> OCLAssembler.OCLBinaryTemplate.PRIVATE_LONG_ARRAY_PTR;
        case Char -> OCLAssembler.OCLBinaryTemplate.PRIVATE_CHAR_ARRAY_PTR;
        case Byte -> OCLAssembler.OCLBinaryTemplate.PRIVATE_BYTE_ARRAY_PTR;
        default -> null;
    };
}
I think can be done with pattern matching for simplicity
```

Comment (mikepapadim) [reply tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/lir/OCLKind.java:276]: not sure if this  suggestion is correct.

File: tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/lir/OCLKind.java [hunk 6]
```diff
@@ -256,6 +256,44 @@ public static OCLAssembler.OCLBinaryTemplate resolvePrivateTemplateType(JavaKind
...
+
+    public static OCLAssembler.OCLBinaryTemplate resolvePrivatePointerCopyTemplate(JavaKind type) {
```
Comment (mikepapadim) [inline tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/lir/OCLKind.java:278]: same here

File: tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/graal/lir/PTXKind.java [hunk 7]
```diff
@@ -250,6 +250,29 @@ public static PTXAssembler.PTXBinaryTemplate resolvePrivateTemplateType(JavaKind
         return null;
     }
 
+    public static PTXAssembler.PTXBinaryTemplate resolvePrivatePointerCopyTemplate(ResolvedJavaType type) {
+        return resolvePrivatePointerCopyTemplate(type.getJavaKind());
+    }
+
+    public static PTXAssembler.PTXBinaryTemplate resolvePrivatePointerCopyTemplate(JavaKind type) {
```
Comment (mikepapadim) [inline tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/graal/lir/PTXKind.java:257]: here maybe also you can use pattern matching

File: tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/graal/nodes/FixedArrayCopyNode.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        this.memoryRegister = memoryRegister;
+        this.pointerCopyTemplate = PTXKind.resolvePrivatePointerCopyTemplate(elementType);
+        this.address = address;
+    }
+
+    public PTXArchitecture.PTXMemoryBase getMemoryRegister() {
+        return memoryRegister;
+    }
+
+    private int getOffset() {
```
Comment (mikepapadim) [inline tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/graal/nodes/FixedArrayCopyNode.java:76]:
```suggestion
    private int getOffset() {
    PTXKind kind = PTXKind.fromResolvedJavaType(elementType);
    return switch (kind) {
        case F64, S64 -> 8;
        case F32, S32 -> 4;
        case S16 -> 2;
        case U16, S8 -> 1;
        default -> -1;
    };
}

```

Comment (mikepapadim) [reply tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/graal/nodes/FixedArrayCopyNode.java:76]: same here

**Review by mikepapadim — APPROVED** [review]
LGTM, thanks @mairooni 

Some minor refactoring to improve code with pattern matching if you have time

**Comment by mairooni** [comment]
The refactoring has been applied, thanks @mikepapadim!

**Comment by jjfumero** [comment]
Great, so we merge this.

**Pull request merged**